cmd/cloud_sql_proxy/build.sh

```diff
@@ -1,39 +1,97 @@
 #!/bin/bash
+#
+# build.sh wraps `go build` to make compiling the Cloud SQL Proxy for
+# distribution more streamlined. When doing normal development on the proxy,
+# `go build .` (or even `go run .`) is sufficient for iterating on the code.
+# This script simply allows a convenient way to cross compile and build a docker
+# container.
+#
+# With no arguments, this script will build a binary marked with "development"
+#
+# Specifying 'release' as the first argument to this script will cross compile
+# for all supported operating systems and architectures. This requires a version
+# identifier to be supplied as the second argument.
+#
+# Specifying 'docker' as the first argument to this script will build a
+# container, tagging it with the second argument. The tag will also be used as
+# the version identifier.
 
 files=$(git status -s)
 if [[ $? != 0 ]]; then
   echo >&2 "Error running git status"
   exit 2
 fi
 
-RELEASE=0
-if [[ "$1" == "release" ]]; then
+# build builds a new binary. Expected variables:
+#   VERSION: string to print out when --version is passed to the final binary
+#   OS: operation system to target (windows, darwin, linux, etc)
+#   ARCH: architecture to target (amd64, 386, etc)
+#   OUT: location to place binary
+build() {
+  echo "   Compile -> $OUT"
+  CGO_ENABLED=0 GOOS=$OS GOARCH=$ARCH go build \
+    -ldflags "-X 'main.versionString=$VERSION'" -a -installsuffix cgo -o $OUT \
+    github.com/GoogleCloudPlatform/cloudsql-proxy/cmd/cloud_sql_proxy
+}
+
+# git_version echos out version information related to the git repo and date.
+git_version() {
+  edits=""
+  if [[ "$files" != "" ]]; then
+    edits=" (local edits)"
+  fi
+  echo "sha $(git rev-parse HEAD)$edits built $(date)"
+}
+
+set -e
+
+case $1 in
+"release")
   if [[ "$files" != "" ]]; then
     echo >&2 "Can't build a release version with local edits; files:"
     echo >&2 "$files"
     exit 1
   fi
   if [[ "$2" == "" ]]; then
-    echo >&2 "Must provide a version number to use as a second parameter"
+    echo >&2 "Must provide a version number to use as the second parameter"
     exit 1
   fi
-  VERSION="version $2"
-  RELEASE=1
-else
-  VERSION="development"
-fi
-
-VERSION+="; sha $(git rev-parse HEAD) built $(date)"
-echo "Compiling $VERSION..."
+  VERSION="version $2; $(git_version)"
+  echo "Cross-compiling $VERSION..."
 
-if [[ $RELEASE == 0 ]]; then
-  CGO_ENABLED=0 go build -ldflags "-X 'main.versionString=$VERSION'" -a -installsuffix cgo -o cloud_sql_proxy .
-else
   for OS in windows darwin linux; do
     for ARCH in amd64 386; do
       OUT="cloud_sql_proxy.$OS.$ARCH"
-      echo "   Compile $OUT"
-      CGO_ENABLED=0 GOOS=$OS GOARCH=$ARCH go build -ldflags "-X 'main.versionString=$VERSION'" -a -installsuffix cgo -o $OUT .
+      build
     done
   done
-fi
+  ;;
+"docker")
+  if [[ "$2" == "" ]]; then
+    echo >&2 "Must provide a version number to use as the second parameter"
```

Comment:
What you're using this for in `docker build` (val to -t) is really the name rather than version number. What you end up with if you pass `1.0` for example is:
```
REPOSITORY                                          TAG                 IMAGE ID            CREATED             SIZE
1.0                                                  latest              26b86071bec3        46 seconds ago      10.75 MB
```

It might be better to ask for both name and version number with version defaulting to the git sha and then use version as the tag. Eg `docker build -t "$2:$(git_version)"`. You can get around this right now by passing a version in the name:tag syntax but it makes some of the echos at least look off. Eg `build.sh docker cloudsql-proxy:e9ce9b2bbef0fbe33bf755b535e049b16513df13` results in: 

```
REPOSITORY                                          TAG                                        IMAGE ID            CREATED             SIZE
cloudsql-proxy         e9ce9b2bbef0fbe33bf755b535e049b16513df13   87a668e55dfa        42 seconds ago      10.75 MB
```